### PR TITLE
Space List and Dashboard Sections Optimisation

### PR DESCRIPTION
**Pull request by kumarpalsinh25**
- Used generic provider param for `SpaceListWidget`
- Optimised code in `dashboard.dart`
- Used generic `EventListWidget` for my upcoming events and my ongoing event view of dashboard
- Used generic `SpaceListWidget` for my spaces and bookmarked spaces section of dashboard

https://github.com/user-attachments/assets/7c787590-1438-4370-9cc4-01cfad4a9726

### CODE REVIEW

File: app/lib/features/home/pages/dashboard.dart [hunk 1]
```diff
@@ -41,34 +39,7 @@ class Dashboard extends ConsumerWidget {
...
-                        const MyEventsSection(
-                          limit: 3,
-                          eventFilters: EventFilters.upcoming,
-                        ),
-                        const MySpacesSection(limit: 5),
-                      ],
-                    )
-                  : emptyState(context),
-            ),
-          ),
```
Comment (kumarpalsinh25) [inline app/lib/features/home/pages/dashboard.dart:71]: Moved this code to separate function.

File: app/lib/features/home/pages/dashboard.dart [hunk 2]
```diff
@@ -106,92 +77,6 @@ class Dashboard extends ConsumerWidget {
...
-              ),
-              const SizedBox(width: 8),
-              Text(title),
-            ],
-          ),
-        ),
-      ),
-    );
-  }
-
```
Comment (kumarpalsinh25) [inline app/lib/features/home/pages/dashboard.dart:194]: Moved this code to separate widget class.

File: app/lib/features/home/pages/dashboard.dart [hunk 3]
```diff
@@ -223,6 +108,27 @@ class Dashboard extends ConsumerWidget {
...
+                  limit: 3,
+                  eventFilters: EventFilters.upcoming,
+                ),
+                MySpacesSection(limit: 5),
+              ],
+            )
+          : emptyState(context),
+    );
+  }
+
```
Comment (kumarpalsinh25) [inline app/lib/features/home/pages/dashboard.dart:131]: Separated dashboard body UI widget code in function.

File: app/lib/features/home/widgets/features_nav_widget.dart [hunk 4]
```diff
@@ -0,0 +1,100 @@
+import 'package:acter/common/themes/colors/color_scheme.dart';
+import 'package:acter/common/utils/routes.dart';
+import 'package:atlas_icons/atlas_icons.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_gen/gen_l10n/l10n.dart';
+import 'package:go_router/go_router.dart';
+
+class FeaturesNavWidget extends StatelessWidget {
```
Comment (kumarpalsinh25) [inline app/lib/features/home/widgets/features_nav_widget.dart:8]: Separated dashboard `FeaturesNavWidget` UI widget code.

File: app/lib/features/home/widgets/my_events.dart [hunk 5]
```diff
@@ -28,66 +19,26 @@ class MyEventsSection extends ConsumerWidget {
...
-    return ListView.builder(
-      shrinkWrap: true,
-      itemCount: count,
-      physics: const NeverScrollableScrollPhysics(),
-      itemBuilder: (context, index) => EventItem(
-        isShowSpaceName: true,
-        margin: const EdgeInsets.only(bottom: 14),
-        event: events[index],
-      ),
+    return EventListWidget(
```
Comment (kumarpalsinh25) [inline app/lib/features/home/widgets/my_events.dart:34]: Used generic `EventListWidget` and removed unwanted duplicated code.

File: app/lib/features/home/widgets/my_spaces_section.dart [hunk 6]
```diff
@@ -25,115 +22,39 @@ class MySpacesSection extends ConsumerWidget {
...
+    final count = limit.map((val) => min(val, allSpacesList.length)) ??
+        allSpacesList.length;
+    return SpaceListWidget(
+      spaceListProvider: spacesProvider,
+      showSectionHeader: true,
+      sectionHeaderTitle: lang.mySpaces,
+      limit: limit,
+      showSectionBg: false,
+      isShowSeeAllButton: count != allSpacesList.length,
+      onClickSectionHeader: () => context.pushNamed(Routes.spaces.name),
```
Comment (kumarpalsinh25) [inline app/lib/features/home/widgets/my_spaces_section.dart:57]: Used generic SpaceListWidget and removed unwanted duplicated code.

File: app/lib/features/home/widgets/my_spaces_section.dart [hunk 7]
```diff
@@ -157,61 +78,34 @@ class _NoSpacesWidgetState extends ConsumerState<_NoSpacesWidget> {
     final lang = L10n.of(context);
     final textTheme = Theme.of(context).textTheme;
     return Column(
+      mainAxisAlignment: MainAxisAlignment.start,
+      crossAxisAlignment: CrossAxisAlignment.stretch,
```
Comment (kumarpalsinh25) [inline app/lib/features/home/widgets/my_spaces_section.dart:82]: Minor UI fixes on `_NoSpacesWidget`

File: app/lib/features/search/pages/quick_search_page.dart [hunk 8]
```diff
@@ -128,8 +129,8 @@ class _QuickSearchPageState extends ConsumerState<QuickSearchPage> {
           if (quickSearchFilters.value == QuickSearchFilters.all ||
               quickSearchFilters.value == QuickSearchFilters.spaces)
             SpaceListWidget(
+              spaceListProvider: spaceListQuickSearchedProvider,
```
Comment (kumarpalsinh25) [inline app/lib/features/search/pages/quick_search_page.dart:132]: `SpaceListWidget` with generic `spaceListProvider`

File: app/lib/features/space/widgets/space_sections/section_header.dart [hunk 9]
```diff
@@ -4,13 +4,15 @@ import 'package:flutter_gen/gen_l10n/l10n.dart';
 
 class SectionHeader extends StatelessWidget {
   final String title;
+  final bool showSectionBg;
```
Comment (kumarpalsinh25) [inline app/lib/features/space/widgets/space_sections/section_header.dart:7]: Enhance `SectionHeader` widget with minor code changes.

File: app/lib/features/spaces/providers/space_list_provider.dart [hunk 10]
```diff
@@ -22,21 +26,48 @@ final subSpacesListProvider =
...
 
-final spaceListSearchProvider = FutureProvider.autoDispose
-    .family<List<Space>, String>((ref, searchText) async {
+final allSpaceListWithBookmarkFrontProvider =
+    Provider.autoDispose<List<Space>>((ref) {
   final bookmarkedSpaceList = ref.watch(bookmarkedSpacesProvider);
   final othersSpaceList = ref.watch(unbookmarkedSpacesProvider);
   final spaceList = bookmarkedSpaceList.followedBy(othersSpaceList);
+  return spaceList.toList();
+});
```
Comment (kumarpalsinh25) [inline app/lib/features/spaces/providers/space_list_provider.dart:35]: All spaces with bookmark front.

File: app/lib/features/spaces/providers/space_list_provider.dart [hunk 11]
```diff
@@ -22,21 +26,48 @@ final subSpacesListProvider =
...
+
+final spaceListSearchProvider = Provider.autoDispose<List<Space>>((ref) {
```
Comment (kumarpalsinh25) [inline app/lib/features/spaces/providers/space_list_provider.dart:37]: Space list provider with search criteria of space list page.

File: app/lib/features/spaces/providers/space_list_provider.dart [hunk 12]
```diff
@@ -22,21 +26,48 @@ final subSpacesListProvider =
...
+    final roomId = space.getRoomIdStr();
+    final spaceInfo = ref.watch(roomAvatarInfoProvider(roomId));
+    final spaceName = spaceInfo.displayName ?? roomId;
+    return spaceName.toLowerCase().contains(searchValue);
+  });
+  return spacesSearchedList.toList();
+});
+
+//Space list for quick search value provider
+final spaceListQuickSearchedProvider = Provider.autoDispose<List<Space>>((ref) {
```
Comment (kumarpalsinh25) [inline app/lib/features/spaces/providers/space_list_provider.dart:57]: Space list provider for quick search page.

File: app/lib/features/spaces/widgets/space_list_widget.dart [hunk 13]
```diff
@@ -1,65 +1,38 @@
...
-      stack: stack,
-      textBuilder: L10n.of(context).loadingFailed,
-      onRetryTap: () {
-        ref.invalidate(spaceListSearchProvider(searchValue ?? ''));
-      },
-    );
-  }
-
-  Widget buildSpaceSectionUI(BuildContext context, List<Space> spaceList) {
+    final spaceList = ref.watch(spaceListProvider);
```
Comment (kumarpalsinh25) [inline app/lib/features/spaces/widgets/space_list_widget.dart:35]: Remove bit of code by converting `spaceListProvider` from `FutureProvider` to simple `Provider`.

File: app/lib/features/spaces/pages/space_list_page.dart [hunk 14]
```diff
@@ -24,15 +23,14 @@ class SpaceListPage extends ConsumerStatefulWidget {
 }
 
 class _AllPinsPageConsumerState extends ConsumerState<SpaceListPage> {
-  String get searchValue => ref.watch(searchValueProvider);
+  String get searchValue => ref.watch(spaceListSearchTermProvider);
```
Comment (gnunicorn) [inline app/lib/features/spaces/pages/space_list_page.dart:26]: I think this isn't needed anymore at all and should be removed.
```suggestion
```

File: app/lib/features/spaces/providers/space_list_provider.dart (hunk 10 above)
Comment (gnunicorn) [reply app/lib/features/spaces/providers/space_list_provider.dart:35]: Let's call that `first` not `front` ...

File: app/lib/features/spaces/providers/space_list_provider.dart [hunk 15]
```diff
@@ -22,21 +26,48 @@ final subSpacesListProvider =
...
+      ref.watch(spaceListSearchTermProvider).trim().toLowerCase();
+
+  //Return all spaces if search is empty
+  final searchValue = searchTerm.trim().toLowerCase();
+  if (searchValue.isEmpty) return spaceList;
+
+  //Return all spaces with search criteria
+  var spacesSearchedList = spaceList.where((space) {
+    final roomId = space.getRoomIdStr();
+    final spaceInfo = ref.watch(roomAvatarInfoProvider(roomId));
```
Comment (gnunicorn) [inline app/lib/features/spaces/providers/space_list_provider.dart:49]: in the past I had noticed that `ref.watch` wouldn't always fully trigger in certain nesting conditions. I believe due them being done in a scope that is lost and thus the ref-watch being removed again. Would be great to have a test for this provider ensure that if the `roomAvatarInfoProvider`-info of an item changes the spaceListSearchProvider properly picks up upon that and shows it in the list.

In the past, this lead to not noticing updates that came in async as being things that are properly matched for the search. It would essentially never refresh displayName-data and if it wasn't available at first try users (in that case back then it was mentions) would never match ...

File: app/lib/features/spaces/providers/space_list_provider.dart [hunk 16]
```diff
@@ -22,21 +26,48 @@ final subSpacesListProvider =
...
 
   //Return all spaces if search is empty
-  final searchValue = searchText.trim().toLowerCase();
-  if (searchValue.isEmpty) return spaceList.toList();
+  final searchValue = searchTerm.trim().toLowerCase();
+  if (searchValue.isEmpty) return spaceList;
 
   //Return all spaces with search criteria
-  return spaceList.where((space) {
+  var spacesSearchedList = spaceList.where((space) {
```
Comment (gnunicorn) [inline app/lib/features/spaces/providers/space_list_provider.dart:66]: var? 
```suggestion
  final spacesSearchedList = spaceList.where((space) {
```

File: app/lib/features/spaces/widgets/space_list_widget.dart [hunk 17]
```diff
@@ -1,65 +1,38 @@
...
 import 'package:flutter/material.dart';
 import 'package:flutter_gen/gen_l10n/l10n.dart';
 import 'package:flutter_riverpod/flutter_riverpod.dart';
-import 'package:logging/logging.dart';
-
-final _log = Logger('a3::space-list-widget');
 
 class SpaceListWidget extends ConsumerWidget {
-  final String? searchValue;
+  final ProviderBase<List<Space>> spaceListProvider;
```
Comment (gnunicorn) [inline app/lib/features/spaces/widgets/space_list_widget.dart:10]: :clap:

**Review by gnunicorn — APPROVED** [review]
lgtm, just a few edge cases I'd like to see tests for to ensure they work as anticipated.
